### PR TITLE
android-interop-testing: include android interop testing in main build

### PR DESCRIPTION
**Pull request by voidzcy**
For #6775.

~TODO: temporarily disabled code shrinking for release buildType, need to fix proguard rules to before being able to do code shrinking.~
The problem was caused by pulling in `org.apache.httpcomponents.htttpcore` by google auth2 lib. According to [Proguard](https://www.guardsquare.com/en/products/proguard/manual/troubleshooting):
> In Android development, sloppy libraries may contain duplicates of classes that are already present in the Android run-time (notably org.w3c.dom, org.xml.sax, org.xmlpull.v1, org.apache.commons.logging.Log, org.apache.http, and org.json). You must remove these classes from your libraries, since they are possibly inconsistent, and the run-time libraries would get precedence anyway.

~TODO: there are tons of errorprone warnings and Proguard notes. Although they do not cause build fail, it would be better to fix/suppress them.~
Proguard notes seem to be fine. There are some compiler warnings coming from javalite generated code, nothing can be done to fix that. A couple of errorprone warnings left, but unclear if it is necessary to fix them.

After the change, generated code will be checked in. So this PR contains generated code. Reviewers should only need to review `.gradle`, `.md` files, `proguard-rules.pro`, `InteropTask.java` and `TesterActivity.java`.

### CODE REVIEW

File: RELEASING.md [hunk 1]
```diff
@@ -32,10 +32,7 @@ them before continuing, and set them again when resuming.
 $ MAJOR=1 MINOR=7 PATCH=0 # Set appropriately for new release
 $ VERSION_FILES=(
   build.gradle
-  android/build.gradle
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply RELEASING.md]: Yep. It's already no longer there. I should have removed this in the change for adding grpc-android into main build (same for grpc-cronet), but forgot to do so. So just do the simple cleanup here.

**Comment by voidzcy** [comment]
> not fully reviewed, sending what i have now
> i ran `./gradlew :grpc-android-interop-testing:{test,checkStyleMain}` those tasks don't work.

You should be able to run `./gradlew :grpc-android-interop-testing:test`. `checkStyle` tasks do not exist in android modules (same for grpc-android and grpc-cronet). They are only added to [modules with `java` plugin](https://github.com/grpc/grpc-java/blob/58e6ad71cc5fcdfab5d523049e925e2f5349aa8b/build.gradle#L232). We could add `checkStyle` tasks (if we are able to) for android modules either manually in each modules buildscript (like what we are currently doing for [javadoc tasks](https://github.com/grpc/grpc-java/blob/master/android/build.gradle#L38)) or find a solution to configure globally in grpc's main buildscript. Anyway, `checkStyle` issue would be addresses separately, together with grpc-android and grpc-cronet.

**Comment by creamsoup** [comment]
> You should be able to run `./gradlew :grpc-android-interop-testing:test`. 

yes, it turns out that it was my local environmental issue.

File: build.gradle [hunk 2]
```diff
@@ -53,6 +54,7 @@ subprojects {
         opencensusVersion = '0.24.0'
 
         configureProtoCompilation = {
+            boolean isAndroid = project.getName().contains('android')
```
Comment (creamsoup) [inline build.gradle]: change seems okay. i am wondering if it is a good idea to split this into two sections; android and java.

Comment (voidzcy) [reply build.gradle]: Configurations for android and java are mostly the same except:
1. Android needs the extra "lite" option for code generation.
2. The build unit for Java project is sourceSet while for Android is variant.

Other than those, they still shares the same configuration in common. I thought about having android's own `configureProtoCompilation` block, but it turned out that majority lines are the same as Java's.

Comment (creamsoup) [reply build.gradle]: okay. yeah little worried about the build becoming more complicated, but splitting sounds like could be harder to maintain

**Comment by dapengzhang0** [comment]
@voidzcy My local run of ` ./gradlew grpc-android-interop-testing:build -PskipAndroid=false` fails as follow

```
> Task :grpc-android-interop-testing:transformClassesAndResourcesWithProguardForRelease
Request to incrementing alive workforce from 0. Current workforce (dead or alive) 0
thread-pool size=4

Note: the configuration refers to the unknown field 'com.google.android.gms.common.api.internal.BasePendingResult$ReleasableResultGuardian mResultGuardian' in class 'com.google.android.gms.common.api.internal.BasePendingResult'
Note: the configuration keeps the entry point 'com.google.android.gms.common.api.internal.LifecycleCallback { com.google.android.gms.common.api.internal.LifecycleFragment getChimeraLifecycleFragmentImpl(com.google.android.gms.common.api.internal.LifecycleActivity); }', but not the descriptor class 'com.google.android.gms.common.api.internal.LifecycleActivity'
Note: the configuration keeps the entry point 'com.google.android.gms.common.images.internal.LoadingImageView { void setOnImageLoadedListener(com.google.android.gms.common.images.ImageManager$OnImageLoadedListener); }', but not the descriptor class 'com.google.android.gms.common.images.ImageManager$OnImageLoadedListener'
Note: the configuration keeps the entry point 'com.google.android.gms.common.images.internal.LoadingImageView { void setClipPathProvider(com.google.android.gms.common.images.internal.LoadingImageView$ClipPathProvider); }', but not the descriptor class 'com.google.android.gms.common.images.internal.LoadingImageView$ClipPathProvider'
Note: there were 1 references to unknown classes.
      You should check your configuration for typos.
      (http://proguard.sourceforge.net/manual/troubleshooting.html#unknownclass)
Note: there were 1 references to unknown class members.
      You should check your configuration for typos.
Note: there were 3 unkept descriptor classes in kept class members.
      You should consider explicitly keeping the mentioned classes
      (using '-keep').
      (http://proguard.sourceforge.net/manual/troubleshooting.html#descriptorclass)
Note: there were 26 unresolved dynamic references to classes or interfaces.
      You should check if you need to specify additional program jars.
      (http://proguard.sourceforge.net/manual/troubleshooting.html#dynamicalclass)
Warning: there were 3 unresolved references to library class members.
         You probably need to update the library versions.
         (http://proguard.sourceforge.net/manual/troubleshooting.html#unresolvedlibraryclassmember)
Warning: Exception while processing task java.io.IOException: Please correct the above warnings first.
Thread(Tasks limiter_1): destruction

> Task :grpc-android-interop-testing:transformClassesAndResourcesWithProguardForRelease FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':grpc-android-interop-testing:transformClassesAndResourcesWithProguardForRelease'.
> Job failed, see logs for details
```

**Comment by voidzcy** [comment]
> @voidzcy My local run of ` ./gradlew grpc-android-interop-testing:build -PskipAndroid=false` fails as follows
> 
> ```
> > Task :grpc-android-interop-testing:transformClassesAndResourcesWithProguardForRelease
> Request to incrementing alive workforce from 0. Current workforce (dead or alive) 0
> thread-pool size=4
> 
> Note: the configuration refers to the unknown field 'com.google.android.gms.common.api.internal.BasePendingResult$ReleasableResultGuardian mResultGuardian' in class 'com.google.android.gms.common.api.internal.BasePendingResult'
> Note: the configuration keeps the entry point 'com.google.android.gms.common.api.internal.LifecycleCallback { com.google.android.gms.common.api.internal.LifecycleFragment getChimeraLifecycleFragmentImpl(com.google.android.gms.common.api.internal.LifecycleActivity); }', but not the descriptor class 'com.google.android.gms.common.api.internal.LifecycleActivity'
> Note: the configuration keeps the entry point 'com.google.android.gms.common.images.internal.LoadingImageView { void setOnImageLoadedListener(com.google.android.gms.common.images.ImageManager$OnImageLoadedListener); }', but not the descriptor class 'com.google.android.gms.common.images.ImageManager$OnImageLoadedListener'
> Note: the configuration keeps the entry point 'com.google.android.gms.common.images.internal.LoadingImageView { void setClipPathProvider(com.google.android.gms.common.images.internal.LoadingImageView$ClipPathProvider); }', but not the descriptor class 'com.google.android.gms.common.images.internal.LoadingImageView$ClipPathProvider'
> Note: there were 1 references to unknown classes.
>       You should check your configuration for typos.
>       (http://proguard.sourceforge.net/manual/troubleshooting.html#unknownclass)
> Note: there were 1 references to unknown class members.
>       You should check your configuration for typos.
> Note: there were 3 unkept descriptor classes in kept class members.
>       You should consider explicitly keeping the mentioned classes
>       (using '-keep').
>       (http://proguard.sourceforge.net/manual/troubleshooting.html#descriptorclass)
> Note: there were 26 unresolved dynamic references to classes or interfaces.
>       You should check if you need to specify additional program jars.
>       (http://proguard.sourceforge.net/manual/troubleshooting.html#dynamicalclass)
> Warning: there were 3 unresolved references to library class members.
>          You probably need to update the library versions.
>          (http://proguard.sourceforge.net/manual/troubleshooting.html#unresolvedlibraryclassmember)
> Warning: Exception while processing task java.io.IOException: Please correct the above warnings first.
> Thread(Tasks limiter_1): destruction
> 
> > Task :grpc-android-interop-testing:transformClassesAndResourcesWithProguardForRelease FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':grpc-android-interop-testing:transformClassesAndResourcesWithProguardForRelease'.
> > Job failed, see logs for details
> ```

Did you pull all commits in this PR? That's caused by ProGuard and I should have fixed it in commit 61a3d75c740197d7896df70cf911e6a70dd72070

**Comment by dapengzhang0** [comment]
Yes, I pulled the PR up to f7c160c84c9ce911b045212c9c04c68e530013b2 . However, I tested build in master and it failed similarly. I figured out the actual failure is

```
Warning: io.grpc.internal.CompositeReadableBuffer$4: can't find referenced method 'java.nio.ByteBuffer limit(int)' in library class java.nio.ByteBuffer
Warning: io.grpc.internal.ReadableBuffers$ByteReadableBufferWrapper: can't find referenced method 'java.nio.ByteBuffer position(int)' in library class java.nio.ByteBuffer
Warning: io.grpc.internal.ReadableBuffers$ByteReadableBufferWrapper: can't find referenced method 'java.nio.ByteBuffer limit(int)' in library class java.nio.ByteBuffer
```

I think it is because I was using jdk 11.  I believe it's not a problem from you change.

**Comment by voidzcy** [comment]
> Yes, I pulled the PR up to [f7c160c](https://github.com/grpc/grpc-java/commit/f7c160c84c9ce911b045212c9c04c68e530013b2) . However, I tested build in master and it failed similarly. I figured out the actual failure is
> 
> ```
> Warning: io.grpc.internal.CompositeReadableBuffer$4: can't find referenced method 'java.nio.ByteBuffer limit(int)' in library class java.nio.ByteBuffer
> Warning: io.grpc.internal.ReadableBuffers$ByteReadableBufferWrapper: can't find referenced method 'java.nio.ByteBuffer position(int)' in library class java.nio.ByteBuffer
> Warning: io.grpc.internal.ReadableBuffers$ByteReadableBufferWrapper: can't find referenced method 'java.nio.ByteBuffer limit(int)' in library class java.nio.ByteBuffer
> ```
> 
> I think it is because I was using jdk 11. I believe it's not a problem from your change.

Yes yes. You need to recompile everything with jdk 8.

**Comment by ericgribkoff** [comment]
> Yes yes. You need to recompile everything with jdk 8.

I'm a little confused about the intended support for JDK versions here. Currently I can build grpc-java at master with JDK 11. With this PR, I need to use JDK 8 (JDK 11 fails) and it sounds like this is intentional. I'm not sure why this would be a desired change, but if it is, should this be explained and documented somewhere?

**Comment by dapengzhang0** [comment]
Java 9 made an ABI incompatible change in `java.nio.ByteBuffer` See https://github.com/apache/felix/pull/114

> I'm a little confused about the intended support for JDK versions here. Currently I can build grpc-java at master with JDK 11. With this PR, I need to use JDK 8 (JDK 11 fails) and it sounds like this is intentional. I'm not sure why this would be a desired change, but if it is, should this be explained and documented somewhere?

**Comment by voidzcy** [comment]
> > Yes yes. You need to recompile everything with jdk 8.
> 
> I'm a little confused about the intended support for JDK versions here. Currently I can build grpc-java at master with JDK 11. With this PR, I need to use JDK 8 (JDK 11 fails) and it sounds like this is intentional. I'm not sure why this would be a desired change, but if it is, should this be explained and documented somewhere?

I suspect this is due to Android. You cannot use JDK 11 to build some part of the source (i.e., `grpc-core`, in which `java.nio.ByteBuffer` is used) and use Android SDK to build `grpc-android-interop-testing` which depends on `grpc-core`. This causes the method not found problem.

**Comment by ericgribkoff** [comment]
My concern was just that a user who tries to build the grpc-java project with JDK 11 will start getting a build error, the cause of which is not very obvious (at least to me, I didn't immediately think JDK incompatibility). Building with `skipAndroid=true` works fine with JDK 11, so maybe that's fine for most people; not sure. But that would mean that my normal workflow would have to change to using JDK 8 or keep using JDK 11 but skipping Android builds whenever I build the project. Both of those are probably fine, but this doesn't seem like it will be at all obvious to newcomers to the project.

**Comment by voidzcy** [comment]
> My concern was just that a user who tries to build the grpc-java project with JDK 11 will start getting a build error, the cause of which is not very obvious (at least to me, I didn't immediately think JDK incompatibility). Building with `skipAndroid=true` works fine with JDK 11, so maybe that's fine for most people; not sure. But that would mean that my normal workflow would have to change to using JDK 8 or keep using JDK 11 but skipping Android builds whenever I build the project. Both of those are probably fine, but this doesn't seem like it will be at all obvious to newcomers to the project.

Alright. This is actually an issue when ProGuard is trying to remove unused code, not really bytecode incompatibility (we have `sourceCompatibility = 1.7` in the [main build file](https://github.com/grpc/grpc-java/blob/master/build.gradle#L233), so real method invocation should be fine). In this case, we could simply add `-dontwarn java.nio.ByteBuffer` to `grpc-android-interop-testing`'s ProGuard rules to ignore that warning. I tried this locally and it seems to work for JDK 11.

**Comment by voidzcy** [comment]
Edited ProGuard rules, build (`skipAndroid=false`) should succeed for JDK 9+.

**Comment by dapengzhang0** [comment]
Good point. I'm trying to fix this in general in #6839 

>My concern was just that a user who tries to build the grpc-java project with JDK 11 will start getting a build error, the cause of which is not very obvious (at least to me, I didn't immediately think JDK incompatibility). Building with skipAndroid=true works fine with JDK 11, so maybe that's fine for most people; not sure. But that would mean that my normal workflow would have to change to using JDK 8 or keep using JDK 11 but skipping Android builds whenever I build the project. Both of those are probably fine, but this doesn't seem like it will be at all obvious to newcomers to the project.

File: build.gradle [hunk 3]
```diff
@@ -1,5 +1,6 @@
 plugins {
-    id "com.android.library" apply false // Necessary for Android plugin to find its classes
+    id "com.android.application" apply false // Necessary for Android plugin to find its classes
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply build.gradle]: The comment also applies to the application plugin.

File: android-interop-testing/proguard-rules.pro [hunk 4]
```diff
@@ -21,3 +21,7 @@
 
 # Ignores: can't find referenced method from grpc-testing's compileOnly dependency on Truth
 -dontwarn io.grpc.testing.DeadlineSubject
+
+# Ignores: can't find referenced method for java.nio.ByteBuffer limit(int) due to JDK ABI
+# incompatibility.
+-dontwarn java.nio.ByteBuffer
```
Comment (dapengzhang0) [inline android-interop-testing/proguard-rules.pro]: This will not be needed once #6839 is in.

Comment (voidzcy) [reply android-interop-testing/proguard-rules.pro]: Yep

Comment (voidzcy) [reply android-interop-testing/proguard-rules.pro]: Reverted.

**Pull request merged**